capirca/lib/versa.py

```diff
@@ -0,0 +1,877 @@
+# Copyright 2011 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Versa generator."""
+# pylint: disable=super-init-not-called
+
+
+import collections
+import copy
+#import datetime
+import itertools
+
+from absl import logging
+from capirca.lib import aclgenerator
+from capirca.lib import nacaddr
+#import six
+
+
+ICMP_TERM_LIMIT = 8
+
+
+class Error(Exception):
+  """generic error class."""
+
+
+class UnsupportedFilterError(Error):
+  """generic error class."""
+  pass
+
+
+class UnsupportedHeaderError(Error):
+  """generic error class."""
+  pass
+
+
+class VersaDuplicateTermError(Error):
+  """generic error class."""
+  pass
+
+class VersaUnsupportedTerm(Error):
+  """generic error class."""
+  pass
+
+class VersaVerbatimError(Error):
+  """generic error class."""
+  pass
+
+
+class VersaOptionError(Error):
+  """generic error class."""
+  pass
+
+
+class MixedAddrBookTypesError(Error):
+  """generic error class."""
+  pass
+
+
+class ConflictingTargetOptionsError(Error):
+  """generic error class."""
+  pass
+
+
+class ConflictingApplicationSetsError(Error):
+  """generic error class."""
+  pass
+
+class Tree:
+  """Creates a Tree Object."""
+  target=[]
+  INDENT = '    '
+
+  def __init__(self, name='root',typ=None):
+    """The init function"""
+    self.children = []
+    self.name = name
+    self.typ = typ
+
+
+  def AddParent(self, parent=None):
+    """Add a Node to Parent"""
+    if isinstance(parent, Tree):
+      parent.AddNode(self)
+
+  def __repr__(self):
+    """repr for a Node """
+    return self.name
+
+  def AddNode(self, node):
+    """add a Node """
+    assert isinstance(node, Tree)
+    self.children.append(node)
+
+  def FindNode(self, nodename):
+    """find a Node """
+    if self.name == nodename:
+      return self
+    if self.children:
+      for child in self.children:
+        ret = child.FindNode(nodename)
+        if isinstance(ret, Tree):
+          return ret
+    return None
+
+  # Print the tree
+  def PrintTree(self,num=0):
+    """Print to the tree.It returns the target """
+    self.ResetTarget()
+    self.PrintTreeInt(num)
+    return self.target
+
+  def PrintTreeInt(self,num=0):
+    """Internal function to the tree. Does recursion"""
```

Comment:
We might need to check if the recursion bounds are sufficient for large policies here.

Reply:
Already verified with multiple tests.